compiler/luci-codegen/README.md

```diff
@@ -0,0 +1,203 @@
+        -DLLVM_ENABLE_TERMINFO=OFF -DLLVM_ENABLE_ASSERTIONS=ON \
+        -DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON -DLLVM_BUILD_32_BITS=OFF \
+        ../llvm
+$ make -j$(nproc)
+$ make install
+$ export LLVM_DIR=${PWD}../install/lib/cmake/llvm
+
+```
+
+##### Halide build
+
+Tested on commit **890a5199ae8ade759de076217670ae980bf44823** Dec 23.
+
+If you did not export variable from previous LLVM build you can pass it via cmake argument: `-DLLVM_DIR=<path to llvm cmake config>` 
+
+```
+$ git clone https://github.com/halide/Halide.git
+$ cd Halide
+$ mkdir build; cd build
+$ cmake -DCMAKE_BUILD_TYPE=Release \
+        -DCMAKE_INSTALL_PREFIX=../install \
+        ../
+$ make -j$(nproc)
+$ make install
+$ export Halide_DIR=${PWD}/../install/lib/cmake/Halide
+```
+
+##### luci-codegen build
+
+Assuming current dir is where build should be done.
+If `LLVM_DIR` and `Halide_DIR` variables are not set you can pass them as cmake arguments.
+
+```
+$ cmake <path to nncc CMakeLists.txt dir>
+$ make -j$(nproc) circle_codegen
+```
+
+`circle_codegen` executable path is `<build dir>/compiler/luci-codegen/src/circle_codegen`
+
+`luci_codegen` library path is `<build dir>/compiler/luci-codegen/src/luci_codegen.so`
+
+## Usage
+
+##### circle_codegen
+
+`luci_codegen` uses Halide autoschedulers, they are built along with main Halide library (libHalide.so on linux)
+and distributed as shared libraries:
+- `libautoschedule_adams2019.so`
+- `libautoschedule_li2018.so`
+- `libautoschedule_mullapudi2016.so`
+
+in order to use `circle_codegen` executable LD_LIBRARY_PATH variable should be set, so these libraries are in path. For example: `$ export LD_PRELOAD_PATH=${Halide_DIR}/../../`.
+
+Basic usage:
+
+```
+$ ./circle_codegen model.circle compiled
+```
+
+As a result there will be created new directory `compiled` where compiler will place all generated files.
+
+This includes:
+
+- `model.circle`: circle model with "compiled" operators replacing parts of graph
+- `generated_subgraph_XX.o`: Halide generated code
+- `generated_subgraph_XX.h`: Halide generated header that describes interfaces of generated code
+- `generated_subgraph_XX.cpp`: Wrappers around Halide generated code. Needed to break dependencies between Halide and runtimes (see section Wrappers for more info)
+
+`generated_subgraph_*` files should be compiled in library and distributed along with `compiled.circle` model.
+
+##### luci_codegen
+
+This is library, and it has simple interface, see `include/Codegen.h` for details.
+Common usage looks like:
+```c++
+  std::unique_ptr<luci::Module> luci_module = get_luci_module_somehow();
+
+// First need to set parameters for compilation.
+// These parameters have default values, so it is optional
+// By default code will be compiled for host architecture
+  luci_codegen::CodegenOptions options;
+  options.max_inline_buffer_threshold = 1024;
+  options.arch.type = luci_codegen::ArchType::Native;
+  options.arch.l1_size = 16*1024;
```

Comment:
AFAIK, Halide gets `last-level cache` size as argument.

It would be better to rename this variable as `last_level_cache_size`.

Reply:
You are right.
This is actually a mistake. For some reason I swapped "last level" and "fist level" caches in my mind =)